syntax = "proto3";

import "lib/relation/proto/query.proto";

package NOrm.NApi;

// Standart values
message TString {
    string value = 1;
}

message TInt {
    int32 value = 1;
}

message TFloat {
    float value = 1;
}

message TBool {
    bool value = 1;
}

message TDefault {}

// Select expressions
message TAll {}

message TColumn {
    repeated uint32 field_path = 1;
    NOrm.NQuery.EColumnType type = 2;
}

message TExpression {
    NOrm.NQuery.EExpressionType expression_type = 1;
    repeated int32 operands = 2;
}

message TSelect {
    uint32 table_num = 1;
    repeated int32 selectors = 2;
    optional int32 where = 3;
    optional int32 group_by = 4;
    optional int32 having = 5;
    optional int32 order_by = 6;
    optional int32 limit = 7;
}

// Set field values
message TUpdateField {
    int32 column_path = 1;
    int32 expression = 2;
}

message TAttribute {
    repeated uint32 path = 1;
    bytes payload = 2;
}

message TInsertSubrequest {
    repeated TAttribute attributes = 1;
}

message TInsert {
    uint32 table_num = 1;
    repeated TInsertSubrequest subrequests = 2;
    bool update_if_exists = 3;
}

message TUpdateSubrequest {
    repeated TAttribute attributes = 1;
}

// Update
message TUpdate {
    uint32 table_num = 1;
    repeated TUpdateSubrequest updates = 2;
}

// Delete
message TDelete {
    uint32 table_num = 1;
    optional int32 where = 2;
}

// Truncate operation
message TTruncate {
    uint32 table_num = 1;
}

// Transaction operations
message TStartTransaction {
    bool read_only = 1;
}

message TCommitTransaction {
}

message TRollbackTransaction {
}

message TClause {
    oneof value {
        TSelect select = 1;
        TInsert insert = 2;
        TUpdate update = 3;
        TDelete delete = 4;
        TTruncate truncate = 5;
        TStartTransaction start_transaction = 9;
        TCommitTransaction commit_transaction = 10;
        TRollbackTransaction rollback_transaction = 11;

        TString string = 21;
        TInt integer = 22;
        TFloat float = 23;
        TBool bool = 24;
        TExpression expression = 25;
        TDefault default = 26;

        TColumn column = 31;
        TAll all = 32;
    }
}

message TQuery {
    repeated TClause clauses = 1;
    repeated int32 start_points = 2;
}
